csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/SourceGenerators/DotnetSourceGeneratorBase.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Semmle.Util;
+using Semmle.Util.Logging;
+
+namespace Semmle.Extraction.CSharp.DependencyFetching
+{
+    internal abstract class DotnetSourceGeneratorBase<T> : SourceGeneratorBase where T : DotnetSourceGeneratorWrapper
+    {
+        protected readonly FileProvider fileProvider;
+        protected readonly FileContent fileContent;
+        protected readonly IDotNet dotnet;
+        protected readonly ICompilationInfoContainer compilationInfoContainer;
+        protected readonly IEnumerable<string> references;
+
+        public DotnetSourceGeneratorBase(
+            FileProvider fileProvider,
+            FileContent fileContent,
+            IDotNet dotnet,
+            ICompilationInfoContainer compilationInfoContainer,
+            ILogger logger,
+            TemporaryDirectory tempWorkingDirectory,
+            IEnumerable<string> references) : base(logger, tempWorkingDirectory)
+        {
+            this.fileProvider = fileProvider;
+            this.fileContent = fileContent;
+            this.dotnet = dotnet;
+            this.compilationInfoContainer = compilationInfoContainer;
+            this.references = references;
+        }
+
+        protected override IEnumerable<string> Run()
+        {
+            if (AdditionalFiles.Count == 0)
+            {
+                logger.LogDebug($"No {FileType} files found. Skipping source generation.");
+                return [];
+            }
+
+            if (!fileContent.IsNewProjectStructureUsed)
+            {
+                logger.LogInfo($"Generating source files from {FileType} files is only supported for new (SDK-style) project files");
+                return [];
+            }
+
+            if (fileProvider.Projects.Count == 0)
+            {
+                logger.LogInfo($"No projects found. Skipping source generation from {FileType} files.");
+                return [];
+            }
+
+            logger.LogInfo($"Generating source files from {AdditionalFiles.Count} {FileType} files...");
+
+            // group additional files by closes project file:
+            var projects = fileProvider.Projects
+                .Select(p => (File: p, Directory: SafeGetDirectoryName(p)))
+                .Where(p => p.Directory.Length > 0);
+
+            var groupedFiles = new Dictionary<string, List<string>>();
+
+            foreach (var additionalFile in AdditionalFiles)
+            {
+                var project = projects
```

Comment:
I don't know many additional files we typically have or how many projects we have - but it sounds like something that can explode for large projects. It is possible to use another data structure to make the "closest" project lookup more efficient.
One suggestion could be to create a `Trie` (tree like structure) based on the projects, where the edges are (sub) folder names. Then the `Trie` can be used to lookup the `closest` by deconstructing the additional file name into a "folder" path.

Reply:
A trie can be implemented using dictionaries (where the keys will be (sub) folder names and the values will be sub-tries.)

Reply:
In case we think this is needed, I am totally up for helping out :-)

Reply:
This is definitely not ideal algorithmically, but I don't think it is causing an issue currently.